src/libs/fileDownload/index.desktop.ts

```diff
@@ -0,0 +1,35 @@
+import ELECTRON_EVENTS from '@desktop/ELECTRON_EVENTS';
+import type Options from '@desktop/electronDownloadManagerType';
+import CONST from '@src/CONST';
+import type {FileDownload} from './types';
+
+/**
+ * The function downloads an attachment on desktop platforms.
+ */
+const fileDownload: FileDownload = (url, fileName) => {
+    const options: Options = {
+        filename: fileName,
+        saveAs: true,
+    };
+    window.electron.send(ELECTRON_EVENTS.DOWNLOAD, {url, options});
+    return new Promise((resolve) => {
+        // This sets a timeout that will resolve the promise after 5 seconds to prevent indefinite hanging
+        const downloadTimeout = setTimeout(() => {
+            resolve();
+        }, CONST.DOWNLOADS_TIMEOUT);
+
+        window.electron.on(ELECTRON_EVENTS.DOWNLOAD_STARTED, (...args: unknown[]) => {
```

Comment:
I think we've discussed this a lot, but I don't think that relying on `DOWNLOAD_STARTED` event would work here. If the queue is already processing a large file, we'll always hit the `downloadTimeout`. A better way would be to listen for a new event called `DOWNLOAD_QUEUED` and emit it whenever the download item is queued.

Reply:
@allroundexperts Another point to consider why it's better to start the loader on onPress instead of `DOWNLOAD_QUEUED` is that the spinner is used to prevent rapid clicks on the same download link. If we start the loader when it's queued, although minimal, it might allow the user to download the same file multiple times.

Have a look at this video:

https://github.com/Expensify/App/assets/11847279/08b482d9-13ef-44bd-8395-ff40d339970c

Please re-consider.

<details><summary>applied changes</summary>

```tsx
diff --git a/desktop/ELECTRON_EVENTS.ts b/desktop/ELECTRON_EVENTS.ts
index cb0b05f097..fd0952b083 100644
--- a/desktop/ELECTRON_EVENTS.ts
+++ b/desktop/ELECTRON_EVENTS.ts
@@ -11,6 +11,7 @@ const ELECTRON_EVENTS = {
     UPDATE_DOWNLOADED: 'update-downloaded',
     DOWNLOAD: 'download',
     DOWNLOAD_STARTED: 'download-started',
+    DOWNLOAD_QUEUED: 'download-queued',
     SILENT_UPDATE: 'silent-update',
 } as const;
 
diff --git a/desktop/contextBridge.ts b/desktop/contextBridge.ts
index e9c61e6953..5aadbc4651 100644
--- a/desktop/contextBridge.ts
+++ b/desktop/contextBridge.ts
@@ -26,6 +26,7 @@ const WHITELIST_CHANNELS_MAIN_TO_RENDERER = [
     ELECTRON_EVENTS.FOCUS,
     ELECTRON_EVENTS.BLUR,
     ELECTRON_EVENTS.DOWNLOAD_STARTED,
+    ELECTRON_EVENTS.DOWNLOAD_QUEUED,
 ] as const;
 
 const getErrorMessage = (channel: string): string => `Electron context bridge cannot be used with channel '${channel}'`;
diff --git a/desktop/createDownloadQueue.ts b/desktop/createDownloadQueue.ts
index 29ce595fc5..c32fa5ad9f 100644
--- a/desktop/createDownloadQueue.ts
+++ b/desktop/createDownloadQueue.ts
@@ -108,6 +108,8 @@ const createDownloadQueue = () => {
 
     const enqueueDownloadItem = (item: DownloadItem): void => {
         queue.enqueue(item);
+        item.win.webContents.send(ELECTRON_EVENTS.DOWNLOAD_QUEUED, {url: item.url});
+
     };
     return {enqueueDownloadItem, dequeueDownloadItem: queue.dequeue};
 };
diff --git a/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx b/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx
index d6bdd241dc..ad51aa8c5d 100644
--- a/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx
+++ b/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx
@@ -48,8 +48,7 @@ function BaseAnchorForAttachmentsOnly({style, source = '', displayName = '', dow
                         if (isDownloading || isOffline) {
                             return;
                         }
-                        Download.setDownload(sourceID, true);
-                        fileDownload(sourceURLWithAuth, displayName, '', Browser.isMobileSafari()).then(() => Download.setDownload(sourceID, false));
+                        fileDownload(sourceURLWithAuth, displayName, '', Browser.isMobileSafari(), () => { Download.setDownload(sourceID, true) }).then(() => Download.setDownload(sourceID, false));
                     }}
                     onPressIn={onPressIn}
                     onPressOut={onPressOut}
diff --git a/src/libs/fileDownload/index.desktop.ts b/src/libs/fileDownload/index.desktop.ts
index 90faa1285b..8e1c213883 100644
--- a/src/libs/fileDownload/index.desktop.ts
+++ b/src/libs/fileDownload/index.desktop.ts
@@ -6,7 +6,7 @@ import type {FileDownload} from './types';
 /**
  * The function downloads an attachment on desktop platforms.
  */
-const fileDownload: FileDownload = (url, fileName) => {
+const fileDownload: FileDownload = (url, fileName, successMessage = '', shouldOpenExternalLink = false, startDownload = () => {}) => {
     const options: Options = {
         filename: fileName,
         saveAs: true,
@@ -18,6 +18,17 @@ const fileDownload: FileDownload = (url, fileName) => {
             resolve();
         }, CONST.DOWNLOADS_TIMEOUT);
 
+        window.electron.on(ELECTRON_EVENTS.DOWNLOAD_QUEUED, (...args: unknown[]) => {
+            const arg = Array.isArray(args) ? args[0] : null;
+            const eventUrl = arg && typeof arg === 'object' && 'url' in arg ? arg.url : null;
+
+            // This event is triggered for all active download instances. We intentionally keep other promises waiting.
+            // Early resolution or rejection of other promises could prematurely stop the loading spinner or prevent the promise from being resolved.
+            if (eventUrl === url) {
+                startDownload();
+            }
+        });
+
         window.electron.on(ELECTRON_EVENTS.DOWNLOAD_STARTED, (...args: unknown[]) => {
             const arg = Array.isArray(args) ? args[0] : null;
             const eventUrl = arg && typeof arg === 'object' && 'url' in arg ? arg.url : null;
diff --git a/src/libs/fileDownload/types.ts b/src/libs/fileDownload/types.ts
index fcc210c1c4..99763dfc42 100644
--- a/src/libs/fileDownload/types.ts
+++ b/src/libs/fileDownload/types.ts
@@ -1,6 +1,6 @@
 import type {Asset} from 'react-native-image-picker';
 
-type FileDownload = (url: string, fileName?: string, successMessage?: string, shouldOpenExternalLink?: boolean) => Promise<void>;
+type FileDownload = (url: string, fileName?: string, successMessage?: string, shouldOpenExternalLink?: boolean, startDownload?: () => void) => Promise<void>;
 
 type ImageResolution = {width: number; height: number};
 type GetImageResolution = (url: File | Asset) => Promise<ImageResolution>;

```

</details>

Reply:
@allroundexperts may I ask why we should initiate the loading spinner within the `DOWNLOAD_QUEUE` instead of placing it in the `onPress` function? What major advantage does this approach offer?

Reply:
I don't think there's any great advantage. It's just that it seems like a more event driven approach. Let's skip that if you feel too strongly about it.

Reply:
New changes have been applied